### PR TITLE
actix-verify-sig

### PR DESCRIPTION
**Pull request by hubenokdev**

### CODE REVIEW

**Comment by hubenokdev** [comment]
Add test function and fix some

File: actix-verify/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+[dependencies]
+# actix-multipart = "0.4"
+actix-web = "4.2.1"
+base64 = "0.20.0"
+secp256k1="0.26.0"
+sha2="0.10.6"
+log ="0.4.17"
+env_logger="0.10.0"
+hex="0.4.3"
```
Comment (ababo) [inline actix-verify/Cargo.toml]: Please consider adding a trailing line to the end of the file.

File: actix-verify/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "actix"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+# actix-multipart = "0.4"
```
Comment (ababo) [inline actix-verify/Cargo.toml]: Please remove all the comments (in all the files) that don't help reader to understand the code.

File: actix-verify/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+mod utils;
+use utils::*;
+use log::{error, info};
+
+#[actix_web::main]
+async fn main() -> Result<(), Error> {
+    std::env::set_var("RUST_LOG", "info,warn,error");
+    env_logger::init();
+
+    println!("Server running on 127.0.0.1:8080");
```
Comment (ababo) [inline actix-verify/src/main.rs]: Please consider make this address configurable.

File: actix-verify/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    HttpServer::new(move || App::new()
+        .wrap(Logger::default())
+        .route("/", web::get().to(index))
+        .route("/verify", web::get().to(verify))
+        )
+        .bind("0.0.0.0:8080")?
```
Comment (ababo) [inline actix-verify/src/main.rs]: This doesn't match to the line logged above.

File: actix-verify/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        .run()
+        .await?;
+
+    Ok(())
+}
+
+pub async fn index() -> HttpResponse {
+    info!("Index Called");
```
Comment (ababo) [inline actix-verify/src/main.rs]: Please don't mix `println!()` vs logging.

File: actix-verify/src/utils.rs [hunk 6]
```diff
@@ -0,0 +1,157 @@
+use actix_web::HttpRequest;
+
+use super::*;
+
+// verification function
```
Comment (ababo) [inline actix-verify/src/utils.rs:8]: Please document public members properly (for cargo doc).

File: actix-verify/src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+pub async fn verify_sig(request: &HttpRequest) -> bool {
+    // Getting the public from headers
+    let hm_key = match request.headers().get("HM-KEY") {
```
Comment (ababo) [inline actix-verify/src/utils.rs]: Consider using `if let Some(pk) = request.headers().get("HM-KEY") ...` instead.

File: actix-verify/src/utils.rs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        Some(pk) => {
+            info!("{:?}", pk);
+            pk.to_str().unwrap()
```
Comment (ababo) [inline actix-verify/src/utils.rs]: So the code will panic if the headers are improperly encoded?

File: actix-verify/src/utils.rs [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        }
+        None => {
+            error!("Public Key not Found");
+            return false;
+        }
+    };
+
+    // Taking the signature from headers
+    let hm_sign = match request.headers().get("HM-SIGN") {
+        Some(sign) => sign.to_str().unwrap(),
```
Comment (ababo) [inline actix-verify/src/utils.rs]: The same is here.

File: actix-verify/src/utils.rs [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        None => {
+            error!("Missing Signature");
+            return false;
+        }
+    };
+
+    // Decoding the timestamp from headers.
+    let ts = match request.headers().get("HM-TS") {
+        Some(timestamp) => timestamp.to_str().unwrap().parse::<u64>().unwrap(),
```
Comment (ababo) [inline actix-verify/src/utils.rs]: We will panic with malformed timestamp?

File: actix-verify/src/utils.rs [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+    let public_key = PublicKey::from_slice(&hex::decode(hm_key).unwrap()).unwrap();
+    info!("Public key  found {}", public_key);
+    // Extract signature from string
+    let signature = Signature::from_compact(&base64::decode(hm_sign).unwrap());
+    // let s :Signature;
+    match signature {
+        Ok(x) => println!("{:?}", x),
+        // s =x;},
+        Err(e) => {
+            println!("{:?}", e)
```
Comment (ababo) [inline actix-verify/src/utils.rs]: You don't exit here, later you unwrap the signature. So we will panic in this case.

File: actix-verify/src/utils.rs [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    ];
+    #[test]
+    async fn test_verify_sig() {
+        let pubkey = PublicKey::from_slice(&PUBKEY).unwrap();
+        let ts = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_secs()
+            .to_string();
+        let msg = format!("{}{}{}", ts, "GET", "/");
```
Comment (ababo) [inline actix-verify/src/utils.rs:111]: Please use `cargo fmt` and `cargo clippy` and make sure the latter doesn't emit warnings.

**Comment by hubenokdev** [comment]
fixed excluding 2 issues.
please explain in another way.

**Comment by hubenokdev** [comment]
add env and change toml

**Comment by hubenokdev** [comment]
fixed 11 errors

**Comment by al-maisan** [comment]
will not be merged -- author refuses to fix issues

**Pull request closed without merging**